### PR TITLE
fix(react): backdrop for inline modal/popover overlay

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

In the React target, inline overlays (`ion-popover` and `ion-modal`) will not define their nested children custom elements (`ion-backdrop`), when in a production tree-shaken build.

Issue Number: #24449

## What is the new behavior?

Inline overlays will share the same custom element definition logic as the regular overlays. This correctly defines `ion-backdrop` for popover/modal inline overlays.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by sean-perkins** [comment]
I'd love to wait until after the New Year to discuss this with the team 🙂 . I could see use deciding not to expose a new public method on the controller API and instead exporting pure constants for just the popover & modal custom element definitions.

File: packages/react/src/components/IonPopover.tsx [hunk 1]
```diff
@@ -1,9 +1,8 @@
-import { JSX } from '@ionic/core/components';
-import { IonPopover as IonPopoverCmp } from '@ionic/core/components/ion-popover.js';
+import { JSX, popoverController } from '@ionic/core/components';
 
 import { createInlineOverlayComponent } from './createInlineOverlayComponent'
 
 export const IonPopover = /*@__PURE__*/ createInlineOverlayComponent<
   JSX.IonPopover,
   HTMLIonPopoverElement
->('ion-popover', IonPopoverCmp);
```
Comment (liamdebeasi) [inline packages/react/src/components/IonPopover.tsx:9]: If you import `defineCustomElements` instead and call that function, it should define both the overlay and the backdrop. See https://github.com/ionic-team/ionic-framework/blob/main/packages/vue/src/components/IonApp.ts#L3-L7 as an example

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

**Comment by sean-perkins** [comment]
Created a dev build of the latest to test with: `6.0.1-dev.1641412400.d3fd57e` & re-verified everything works as expected with the reproduction app: https://github.com/JohnDeved/ionic-popover-bug.

File: core/src/utils/overlays.ts [hunk 2]
```diff
@@ -59,27 +59,33 @@ export const prepareOverlay = <T extends HTMLIonOverlayElement>(el: T) => {
...
-      /**
-       * If the parent element has nested usage of custom elements,
-       * we need to manually define those custom elements.
-       */
-      if (childrenCustomElements) {
-        for (const customElementDefinition of childrenCustomElements) {
-          if (!window.customElements.get(customElementDefinition.tagName)) {
-            window.customElements.define(customElementDefinition.tagName, customElementDefinition.customElement);
-          }
+  if (typeof window !== 'undefined' && typeof window.customElements !== 'undefined') {
```
Comment (liamdebeasi) [inline core/src/utils/overlays.ts:64]: This line is redundant. If the window SSR check in `createOverlay` fails, then this code won't get run.

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
The original issue says it impacts Vue as well, so we should make this change in https://github.com/ionic-team/ionic-framework/blob/main/packages/vue/src/components/Overlays.ts#L12-L18 as well

Also, my understanding of the issue is this impacts all overlay components, not just modal and popover. If that is the case, components like action sheet also need to be updated: https://github.com/ionic-team/ionic-framework/blob/main/packages/react/src/components/IonActionSheet.tsx#L6

**Comment by sean-perkins** [comment]
Alright this should look better now. Didn't realize the overlays generation script was local to this repo. Simplified not needing to update the output targets. 

Kept the commits to follow the changes, but will squash and rework to better explain it affects react & vue targets.

**Review by liamdebeasi — APPROVED** [review]
Great job!

**Pull request merged**